File: typescript/src/flags/server/provider.ts
import type {
  Provider,
  ResolutionDetails,
  EvaluationContext,
  JsonValue,
  ProviderMetadata,
  Hook,
  ProviderStatus,
  ErrorCode,
  FlagMetadata,
} from '@openfeature/server-sdk';
import { FlagsClient } from '../client.js';
import type { EvaluationContext as KitbaseContext } from '../types.js';

/**
 * Configuration options for KitbaseProvider (Server)
 */
export interface KitbaseProviderOptions {
  /**
   * Your Kitbase API key
   */
  token: string;

  /**
   * Enable flag caching (default: true)
   */
  cache?: boolean;

  /**
   * Cache TTL in milliseconds (default: 60000 = 1 minute)
   */
  cacheTtl?: number;
}

interface CacheEntry {
  value: unknown;
  variant?: string;
  reason: string;
  errorCode?: ErrorCode;
  errorMessage?: string;
  flagMetadata?: FlagMetadata;
  valueType: string;
  timestamp: number;
}

/**
 * Kitbase OpenFeature Provider for Server-side applications
 *
 * Implements the official OpenFeature Provider interface from @openfeature/server-sdk
 * for seamless integration with the OpenFeature ecosystem.
 *
 * @example
 * ```typescript
 * import { OpenFeature } from '@openfeature/server-sdk';
 * import { KitbaseProvider } from '@kitbase/sdk/flags/server';
 *
 * // Register the provider
 * await OpenFeature.setProviderAndWait(new KitbaseProvider({
 *   token: 'YOUR_API_KEY'
 * }));
 *
 * // Get a client and evaluate flags
 * const client = OpenFeature.getClient();
 * const isEnabled = await client.getBooleanValue('dark-mode', false, {
 *   targetingKey: 'user-123'
 * });
 * ```
 */
export class KitbaseProvider implements Provider {
  readonly metadata: ProviderMetadata = {
    name: 'kitbase',
  };

  readonly rulesChanged?: () => void;
  readonly hooks?: Hook[];
  status?: ProviderStatus;

  private readonly client: FlagsClient;
  private readonly cacheEnabled: boolean;
  private readonly cacheTtl: number;
  private cache: Map<string, CacheEntry> = new Map();

  constructor(options: KitbaseProviderOptions) {
    this.client = new FlagsClient({ token: options.token });
    this.cacheEnabled = options.cache ?? true;
    this.cacheTtl = options.cacheTtl ?? 60000;
    this.status = 'NOT_READY' as ProviderStatus;
  }

  /**
   * Initialize the provider
   */
  async initialize(): Promise<void> {
    // Provider is ready immediately - no async setup needed
    this.status = 'READY' as ProviderStatus;
  }

  /**
   * Shutdown the provider
   */
  async onClose(): Promise<void> {
    this.cache.clear();
    this.status = 'NOT_READY' as ProviderStatus;
  }

  /**
   * Resolve a boolean flag evaluation
   */
  async resolveBooleanEvaluation(
    flagKey: string,
    defaultValue: boolean,
    context: EvaluationContext,
  ): Promise<ResolutionDetails<boolean>> {
    return this.resolveEvaluation(flagKey, defaultValue, context, 'boolean');
  }

  /**
   * Resolve a string flag evaluation
   */
  async resolveStringEvaluation(
    flagKey: string,
    defaultValue: string,
    context: EvaluationContext,
  ): Promise<ResolutionDetails<string>> {
    return this.resolveEvaluation(flagKey, defaultValue, context, 'string');
  }

  /**
   * Resolve a number flag evaluation
   */
  async resolveNumberEvaluation(
    flagKey: string,
    defaultValue: number,
    context: EvaluationContext,
  ): Promise<ResolutionDetails<number>> {
    return this.resolveEvaluation(flagKey, defaultValue, context, 'number');
  }

  /**
   * Resolve an object flag evaluation
   */
  async resolveObjectEvaluation<T extends JsonValue>(
    flagKey: string,
    defaultValue: T,
    context: EvaluationContext,
  ): Promise<ResolutionDetails<T>> {
    return this.resolveEvaluation(flagKey, defaultValue, context, 'json');
  }

  /**
   * Prefetch all flags for a context (useful for caching)
   */
  async prefetchFlags(context?: EvaluationContext): Promise<void> {
    const kitbaseContext = this.toKitbaseContext(context);
    const snapshot = await this.client.getSnapshot({ context: kitbaseContext });

    for (const flag of snapshot.flags) {
      const cacheKey = this.getCacheKey(flag.flagKey, context);
      this.cache.set(cacheKey, {
        value: flag.value,
        variant: flag.variant,
        reason: flag.reason,
        errorCode: flag.errorCode as ErrorCode | undefined,
        errorMessage: flag.errorMessage,
        flagMetadata: flag.flagMetadata as FlagMetadata | undefined,
        valueType: flag.valueType,
        timestamp: Date.now(),
      });
    }
  }

  /**
   * Clear the flag cache
   */
  clearCache(): void {
    this.cache.clear();
  }

  /**
   * Generic evaluation resolution
   */
  private async resolveEvaluation<T>(
    flagKey: string,
    defaultValue: T,
    context: EvaluationContext,
    expectedType: 'boolean' | 'string' | 'number' | 'json',
  ): Promise<ResolutionDetails<T>> {
    // Check cache first
    if (this.cacheEnabled) {
      const cached = this.getCachedValue<T>(flagKey, context, expectedType);
      if (cached) {
        return cached;
      }
    }

    try {
      const kitbaseContext = this.toKitbaseContext(context);
      const flag = await this.client.evaluateFlag(flagKey, {
        context: kitbaseContext,
        defaultValue,
      });

      // Check for type mismatch
      if (flag.valueType !== expectedType) {
        return {
          value: defaultValue,
          reason: 'ERROR',
          errorCode: 'TYPE_MISMATCH' as ErrorCode,
          errorMessage: `Expected ${expectedType}, got ${flag.valueType}`,
        };
      }

      // Cache the result
      if (this.cacheEnabled) {
        this.setCachedValue(flagKey, context, flag, expectedType);
      }

      // Handle disabled flags or errors
      if (
        flag.errorCode === 'FLAG_NOT_FOUND' ||
        !flag.enabled ||
        flag.value === null ||
        flag.value === undefined
      ) {
        return {
          value: defaultValue,
          variant: flag.variant,
          reason: flag.reason,
          errorCode: flag.errorCode as ErrorCode | undefined,
          errorMessage: flag.errorMessage,
          flagMetadata: flag.flagMetadata as FlagMetadata | undefined,
        };
      }

      return {
        value: flag.value as T,
        variant: flag.variant,
        reason: flag.reason,
        errorCode: flag.errorCode as ErrorCode | undefined,
        errorMessage: flag.errorMessage,
        flagMetadata: flag.flagMetadata as FlagMetadata | undefined,
      };
    } catch (error) {
      // Return default on error (OpenFeature behavior)
      return {
        value: defaultValue,
        reason: 'ERROR',
        errorCode: 'GENERAL' as ErrorCode,
        errorMessage: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  /**
   * Convert OpenFeature context to Kitbase context
   */
  private toKitbaseContext(
    context?: EvaluationContext,
  ): KitbaseContext | undefined {
    if (!context) {
      return undefined;
    }
    return context as KitbaseContext;
  }

  /**
   * Get cache key for a flag + context combination
   */
  private getCacheKey(flagKey: string, context?: EvaluationContext): string {
    const contextKey = context?.targetingKey ?? 'anonymous';
    return `${flagKey}:${contextKey}`;
  }

  /**
   * Get cached value if valid
   */
  private getCachedValue<T>(
    flagKey: string,
    context: EvaluationContext | undefined,
    expectedType: string,
  ): ResolutionDetails<T> | null {
    const cacheKey = this.getCacheKey(flagKey, context);
    const cached = this.cache.get(cacheKey);

    if (!cached) {
      return null;
    }

    // Check if cache is expired
    if (Date.now() - cached.timestamp > this.cacheTtl) {
      this.cache.delete(cacheKey);
      return null;
    }

    // Check type match
    if (cached.valueType !== expectedType) {
      return null;
    }

    return {
      value: cached.value as T,
      variant: cached.variant,
      reason: 'CACHED',
      errorCode: cached.errorCode,
      errorMessage: cached.errorMessage,
      flagMetadata: cached.flagMetadata,
    };
  }

  /**
   * Set cached value
   */
  private setCachedValue(
    flagKey: string,
    context: EvaluationContext | undefined,
    flag: {
      value: unknown;
      variant?: string;
      reason: string;
      errorCode?: string;
      errorMessage?: string;
      flagMetadata?: Record<string, unknown>;
      valueType: string;
    },
    valueType: string,
  ): void {
    const cacheKey = this.getCacheKey(flagKey, context);
    this.cache.set(cacheKey, {
      value: flag.value,
      variant: flag.variant,
      reason: flag.reason,
      errorCode: flag.errorCode as ErrorCode | undefined,
      errorMessage: flag.errorMessage,
      flagMetadata: flag.flagMetadata as FlagMetadata | undefined,
      valueType,
      timestamp: Date.now(),
    });
  }
}

File: flags/dart/lib/src/openfeature/provider.dart
import 'dart:async';

import 'package:kitbase_flags/client.dart';
import 'package:kitbase_flags/types.dart' as kitbase_types;
import 'package:openfeature_dart_server_sdk/feature_provider.dart';

/// Configuration options for KitbaseProvider
class KitbaseProviderOptions {
  /// Your Kitbase API key
  final String token;

  /// Enable flag caching (default: true)
  final bool cache;

  /// Cache TTL in milliseconds (default: 60000 = 1 minute)
  final int cacheTtl;

  const KitbaseProviderOptions({
    required this.token,
    this.cache = true,
    this.cacheTtl = 60000,
  });
}

class _CacheEntry {
  final dynamic value;
  final String? variant;
  final String reason;
  final String? errorCode;
  final String? errorMessage;
  final Map<String, dynamic>? flagMetadata;
  final String valueType;
  final DateTime timestamp;

  _CacheEntry({
    required this.value,
    this.variant,
    required this.reason,
    this.errorCode,
    this.errorMessage,
    this.flagMetadata,
    required this.valueType,
    required this.timestamp,
  });
}

/// Kitbase OpenFeature Provider for server-side applications.
///
/// Implements the official OpenFeature Provider interface from
/// `openfeature_dart_server_sdk` for seamless integration with
/// the OpenFeature ecosystem.
///
/// ```dart
/// import 'package:openfeature_dart_server_sdk/open_feature_api.dart';
/// import 'package:openfeature_dart_server_sdk/client.dart';
/// import 'package:openfeature_dart_server_sdk/hooks.dart';
/// import 'package:openfeature_dart_server_sdk/evaluation_context.dart';
/// import 'package:kitbase/flags/openfeature.dart';
///
/// void main() async {
///   // Register the Kitbase provider
///   final api = OpenFeatureAPI();
///   api.setProvider(KitbaseProvider(
///     options: KitbaseProviderOptions(token: 'YOUR_API_KEY'),
///   ));
///
///   // Create a client
///   final client = FeatureClient(
///     metadata: ClientMetadata(name: 'my-app'),
///     hookManager: HookManager(),
///     defaultContext: EvaluationContext(attributes: {}),
///   );
///
///   // Evaluate flags
///   final isEnabled = await client.getBooleanFlag(
///     'dark-mode',
///     defaultValue: false,
///     context: EvaluationContext(attributes: {
///       'targetingKey': 'user-123',
///       'plan': 'premium',
///     }),
///   );
///
///   print('Dark mode enabled: $isEnabled');
/// }
/// ```
class KitbaseProvider implements FeatureProvider {
  @override
  String get name => 'kitbase';

  @override
  ProviderMetadata get metadata => ProviderMetadata(name: name);

  @override
  ProviderState get state => _state;
  ProviderState _state = ProviderState.NOT_READY;

  @override
  ProviderConfig get config => const ProviderConfig();

  final KitbaseFlags _client;
  final bool _cacheEnabled;
  final int _cacheTtl;
  final Map<String, _CacheEntry> _cache = {};

  /// Creates a new KitbaseProvider.
  ///
  /// [options] contains the configuration including your API token.
  KitbaseProvider({required KitbaseProviderOptions options})
      : _client = KitbaseFlags(token: options.token),
        _cacheEnabled = options.cache,
        _cacheTtl = options.cacheTtl;

  @override
  Future<void> initialize([Map<String, dynamic>? config]) async {
    _state = ProviderState.READY;
  }

  @override
  Future<void> connect() async {
    // No persistent connection needed
  }

  @override
  Future<void> shutdown() async {
    _cache.clear();
    _client.close();
    _state = ProviderState.NOT_READY;
  }

  @override
  Future<FlagEvaluationResult<bool>> getBooleanFlag(
    String flagKey,
    bool defaultValue, {
    Map<String, dynamic>? context,
  }) async {
    return _evaluateFlag<bool>(flagKey, defaultValue, context, 'boolean');
  }

  @override
  Future<FlagEvaluationResult<String>> getStringFlag(
    String flagKey,
    String defaultValue, {
    Map<String, dynamic>? context,
  }) async {
    return _evaluateFlag<String>(flagKey, defaultValue, context, 'string');
  }

  @override
  Future<FlagEvaluationResult<int>> getIntegerFlag(
    String flagKey,
    int defaultValue, {
    Map<String, dynamic>? context,
  }) async {
    final result =
        await _evaluateFlag<num>(flagKey, defaultValue, context, 'number');
    return FlagEvaluationResult<int>(
      flagKey: result.flagKey,
      value: result.value.toInt(),
      evaluatedAt: result.evaluatedAt,
      evaluatorId: result.evaluatorId,
      variant: result.variant,
      reason: result.reason,
      errorCode: result.errorCode,
      errorMessage: result.errorMessage,
    );
  }

  @override
  Future<FlagEvaluationResult<double>> getDoubleFlag(
    String flagKey,
    double defaultValue, {
    Map<String, dynamic>? context,
  }) async {
    final result =
        await _evaluateFlag<num>(flagKey, defaultValue, context, 'number');
    return FlagEvaluationResult<double>(
      flagKey: result.flagKey,
      value: result.value.toDouble(),
      evaluatedAt: result.evaluatedAt,
      evaluatorId: result.evaluatorId,
      variant: result.variant,
      reason: result.reason,
      errorCode: result.errorCode,
      errorMessage: result.errorMessage,
    );
  }

  @override
  Future<FlagEvaluationResult<Map<String, dynamic>>> getObjectFlag(
    String flagKey,
    Map<String, dynamic> defaultValue, {
    Map<String, dynamic>? context,
  }) async {
    return _evaluateFlag<Map<String, dynamic>>(
        flagKey, defaultValue, context, 'json');
  }

  /// Prefetch all flags for a context (useful for caching).
  Future<void> prefetchFlags({Map<String, dynamic>? context}) async {
    final kitbaseContext = _toKitbaseContext(context);
    final snapshot = await _client.getSnapshot(
        options: kitbase_types.EvaluateOptions(context: kitbaseContext));

    for (final flag in snapshot.flags) {
      final cacheKey = _getCacheKey(flag.flagKey, context);
      _cache[cacheKey] = _CacheEntry(
        value: flag.value,
        variant: flag.variant,
        reason: _reasonToString(flag.reason),
        errorCode: flag.errorCode != null
            ? _kitbaseErrorCodeToString(flag.errorCode!)
            : null,
        errorMessage: flag.errorMessage,
        flagMetadata: flag.flagMetadata,
        valueType: kitbase_types.flagValueTypeToString(flag.valueType),
        timestamp: DateTime.now(),
      );
    }
  }

  /// Clear the flag cache.
  void clearCache() {
    _cache.clear();
  }

  Future<FlagEvaluationResult<T>> _evaluateFlag<T>(
    String flagKey,
    T defaultValue,
    Map<String, dynamic>? context,
    String expectedType,
  ) async {
    // Check cache first
    if (_cacheEnabled) {
      final cached = _getCachedValue<T>(flagKey, context, expectedType);
      if (cached != null) {
        return cached;
      }
    }

    try {
      final kitbaseContext = _toKitbaseContext(context);
      final flag = await _client.evaluateFlag(
        flagKey,
        context: kitbaseContext,
        defaultValue: defaultValue,
      );

      // Check for type mismatch
      if (kitbase_types.flagValueTypeToString(flag.valueType) != expectedType) {
        return FlagEvaluationResult<T>(
          flagKey: flagKey,
          value: defaultValue,
          evaluatedAt: DateTime.now(),
          evaluatorId: name,
          reason: 'ERROR',
          errorCode: ErrorCode.TYPE_MISMATCH,
          errorMessage:
              'Expected $expectedType, got ${kitbase_types.flagValueTypeToString(flag.valueType)}',
        );
      }

      // Cache the result
      if (_cacheEnabled) {
        _setCachedValue(flagKey, context, flag, expectedType);
      }

      // Handle disabled flags or errors
      if (flag.errorCode == kitbase_types.ErrorCode.flagNotFound ||
          !flag.enabled ||
          flag.value == null) {
        return FlagEvaluationResult<T>(
          flagKey: flagKey,
          value: defaultValue,
          evaluatedAt: DateTime.now(),
          evaluatorId: name,
          variant: flag.variant,
          reason: _reasonToString(flag.reason),
          errorCode: flag.errorCode != null
              ? _kitbaseErrorCodeToOpenFeatureErrorCode(flag.errorCode!)
              : null,
          errorMessage: flag.errorMessage,
        );
      }

      return FlagEvaluationResult<T>(
        flagKey: flagKey,
        value: flag.value as T,
        evaluatedAt: DateTime.now(),
        evaluatorId: name,
        variant: flag.variant,
        reason: _reasonToString(flag.reason),
        errorCode: flag.errorCode != null
            ? _kitbaseErrorCodeToOpenFeatureErrorCode(flag.errorCode!)
            : null,
        errorMessage: flag.errorMessage,
      );
    } catch (error) {
      return FlagEvaluationResult<T>(
        flagKey: flagKey,
        value: defaultValue,
        evaluatedAt: DateTime.now(),
        evaluatorId: name,
        reason: 'ERROR',
        errorCode: ErrorCode.GENERAL,
        errorMessage: error.toString(),
      );
    }
  }

  kitbase_types.EvaluationContext? _toKitbaseContext(
      Map<String, dynamic>? context) {
    if (context == null || context.isEmpty) {
      return null;
    }

    final targetingKey = context['targetingKey'] as String?;
    final attributes = Map<String, dynamic>.from(context)
      ..remove('targetingKey');

    return kitbase_types.EvaluationContext(
      targetingKey: targetingKey,
      attributes: attributes,
    );
  }

  String _getCacheKey(String flagKey, Map<String, dynamic>? context) {
    final contextKey = context?['targetingKey'] ?? 'anonymous';
    return '$flagKey:$contextKey';
  }

  FlagEvaluationResult<T>? _getCachedValue<T>(
    String flagKey,
    Map<String, dynamic>? context,
    String expectedType,
  ) {
    final cacheKey = _getCacheKey(flagKey, context);
    final cached = _cache[cacheKey];

    if (cached == null) {
      return null;
    }

    // Check if cache is expired
    if (DateTime.now().difference(cached.timestamp).inMilliseconds >
        _cacheTtl) {
      _cache.remove(cacheKey);
      return null;
    }

    // Check type match
    if (cached.valueType != expectedType) {
      return null;
    }

    return FlagEvaluationResult<T>(
      flagKey: flagKey,
      value: cached.value as T,
      evaluatedAt: cached.timestamp,
      evaluatorId: name,
      variant: cached.variant,
      reason: 'CACHED',
      errorCode: cached.errorCode != null
          ? _stringToOpenFeatureErrorCode(cached.errorCode!)
          : null,
      errorMessage: cached.errorMessage,
    );
  }

  void _setCachedValue(
    String flagKey,
    Map<String, dynamic>? context,
    kitbase_types.EvaluatedFlag flag,
    String valueType,
  ) {
    final cacheKey = _getCacheKey(flagKey, context);
    _cache[cacheKey] = _CacheEntry(
      value: flag.value,
      variant: flag.variant,
      reason: _reasonToString(flag.reason),
      errorCode: flag.errorCode != null
          ? _kitbaseErrorCodeToString(flag.errorCode!)
          : null,
      errorMessage: flag.errorMessage,
      flagMetadata: flag.flagMetadata,
      valueType: valueType,
      timestamp: DateTime.now(),
    );
  }

  String _reasonToString(kitbase_types.ResolutionReason reason) {
    switch (reason) {
      case kitbase_types.ResolutionReason.static_:
        return 'STATIC';
      case kitbase_types.ResolutionReason.default_:
        return 'DEFAULT';
      case kitbase_types.ResolutionReason.targetingMatch:
        return 'TARGETING_MATCH';
      case kitbase_types.ResolutionReason.split:
        return 'SPLIT';
      case kitbase_types.ResolutionReason.cached:
        return 'CACHED';
      case kitbase_types.ResolutionReason.disabled:
        return 'DISABLED';
      case kitbase_types.ResolutionReason.unknown:
        return 'UNKNOWN';
      case kitbase_types.ResolutionReason.stale:
        return 'STALE';
      case kitbase_types.ResolutionReason.error:
        return 'ERROR';
    }
  }

  String _kitbaseErrorCodeToString(kitbase_types.ErrorCode code) {
    switch (code) {
      case kitbase_types.ErrorCode.providerNotReady:
        return 'PROVIDER_NOT_READY';
      case kitbase_types.ErrorCode.flagNotFound:
        return 'FLAG_NOT_FOUND';
      case kitbase_types.ErrorCode.parseError:
        return 'PARSE_ERROR';
      case kitbase_types.ErrorCode.typeMismatch:
        return 'TYPE_MISMATCH';
      case kitbase_types.ErrorCode.targetingKeyMissing:
        return 'TARGETING_KEY_MISSING';
      case kitbase_types.ErrorCode.invalidContext:
        return 'INVALID_CONTEXT';
      case kitbase_types.ErrorCode.general:
        return 'GENERAL';
    }
  }

  /// Convert Kitbase ErrorCode to OpenFeature ErrorCode enum.
  ErrorCode _kitbaseErrorCodeToOpenFeatureErrorCode(
      kitbase_types.ErrorCode code) {
    switch (code) {
      case kitbase_types.ErrorCode.providerNotReady:
        return ErrorCode.PROVIDER_NOT_READY;
      case kitbase_types.ErrorCode.flagNotFound:
        return ErrorCode.FLAG_NOT_FOUND;
      case kitbase_types.ErrorCode.parseError:
        return ErrorCode.PARSE_ERROR;
      case kitbase_types.ErrorCode.typeMismatch:
        return ErrorCode.TYPE_MISMATCH;
      case kitbase_types.ErrorCode.targetingKeyMissing:
        return ErrorCode.TARGETING_KEY_MISSING;
      case kitbase_types.ErrorCode.invalidContext:
        return ErrorCode.INVALID_CONTEXT;
      case kitbase_types.ErrorCode.general:
        return ErrorCode.GENERAL;
    }
  }

  ErrorCode _stringToOpenFeatureErrorCode(String code) {
    switch (code) {
      case 'PROVIDER_NOT_READY':
        return ErrorCode.PROVIDER_NOT_READY;
      case 'FLAG_NOT_FOUND':
        return ErrorCode.FLAG_NOT_FOUND;
      case 'PARSE_ERROR':
        return ErrorCode.PARSE_ERROR;
      case 'TYPE_MISMATCH':
        return ErrorCode.TYPE_MISMATCH;
      case 'TARGETING_KEY_MISSING':
        return ErrorCode.TARGETING_KEY_MISSING;
      case 'INVALID_CONTEXT':
        return ErrorCode.INVALID_CONTEXT;
      case 'GENERAL':
        return ErrorCode.GENERAL;
      default:
        return ErrorCode.GENERAL;
    }
  }
}
